core/drivers/hisi_trng.c

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (c) 2021 HiSilicon Limited. */
+
+#include <initcall.h>
+#include <io.h>
+#include <malloc.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <rng_support.h>
+
+#define HTRNG_RANDATA_REG 0xF0
+#define HTRNG_BYTES 4
+
+#define POLL_PERIOD 10
+#define POLL_TIMEOUT 1000
+
+struct hisi_trng {
+	vaddr_t base;
+	uint32_t ver;
+};
+
+union trng_fifo {
+	uint32_t val;
+	uint8_t data_byte[HTRNG_BYTES];
+};
+
+static struct hisi_trng *trng_dev;
+
+TEE_Result hw_get_random_bytes(void *buf, size_t len)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	uint8_t *out = (uint8_t *)buf;
+	static union trng_fifo fifo = { };
+	size_t buf_pos = 0;
+	size_t fifo_pos = 0;
+
+	if (!trng_dev) {
+		EMSG("No valid TRNG device.");
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	if (!buf || !len) {
+		EMSG("Invalid input parameter.");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	while (buf_pos < len) {
+		if (fifo_pos == 0) {
+			if (IO_READ32_POLL_TIMEOUT(trng_dev->base +
```

Comment:
Doesn't this access need to be serialized in case `hw_get_random_bytes()` called by multiple cores at the same time?